### PR TITLE
reserved token distro tries to pay project

### PR DESCRIPTION
**Pull request by mejango**
# Description

Reserved token's that are split to a projectId will be paid to the project instead of being sent to the project owner.

## Limitations & risks

*Are there any trade-off or new vulnarbility surface based on theses changes?*

# Check-list
- [ ] Tests are covering the new feature
- [ ] Code is [natspec'd](https://docs.soliditylang.org/en/v0.8.17/natspec-format.html)
- [ ] Code is [linted and formatted](https://docs.soliditylang.org/en/v0.8.17/style-guide.html)
- [ ] I have run the test locally (and they pass)
- [ ] I have rebased to the latest main commit (and tests still pass)

# Interactions
These changes will impact the following contracts:
- Directly:

- Indirectly:

### CODE REVIEW

File: src/JBController.sol [hunk 1]
```diff
@@ -803,46 +808,92 @@ contract JBController is JBPermissioned, ERC2771Context, ERC165, IJBController,
...
+                            // Mint the tokens.
+                            TOKENS.mintFor(beneficiary, projectId, splitAmount);
+                        } else {
+                            // Mint the tokens to this contract.
+                            TOKENS.mintFor(address(0), projectId, splitAmount);
+
+                            // Send the projectId in the metadata.
+                            bytes memory metadata = bytes(abi.encodePacked(projectId));
+
+                            // Try to fulfill the payment
```
Comment (drgorillamd) [inline src/JBController.sol:870]: shouldn't we `TOKENS.approve(terminal, splitAmount)` here?

File: src/JBController.sol [hunk 2]
```diff
@@ -803,46 +808,92 @@ contract JBController is JBPermissioned, ERC2771Context, ERC165, IJBController,
...
+
+                        // If the paying project doesn't have a token or the receiving project isn't accepting the
+                        // token,
+                        // fallback to sending to the beneficiary.
+                        if (address(token) == address(0) || address(terminal) == address(0)) {
+                            // Mint the tokens.
+                            TOKENS.mintFor(beneficiary, projectId, splitAmount);
+                        } else {
+                            // Mint the tokens to this contract.
+                            TOKENS.mintFor(address(0), projectId, splitAmount);
```
Comment (drgorillamd) [inline src/JBController.sol:865]: why the address(0)?

File: src/JBController.sol (hunk 1 above)
Comment (mejango) [reply src/JBController.sol:873]: yes!

File: src/JBController.sol (hunk 2 above)
Comment (mejango) [reply src/JBController.sol]: address(this), my bad

File: src/JBController.sol [hunk 3]
```diff
@@ -803,46 +808,98 @@ contract JBController is JBPermissioned, ERC2771Context, ERC165, IJBController,
...
                         })
                     );
+                    // If there's a project ID, try to pay the project. If it fails, fallback to paying the beneficiary.
+                } else {
+                    // Use the split's beneficiary if provided, otherwise use the msg sender.
+                    address beneficiary = split.beneficiary != address(0) ? split.beneficiary : _msgSender();
+
+                    if (split.projectId != 0) {
+                        // Get a reference to the project's token. This will return the 0 address if the project doesn't
+                        // yet
```
Comment (drgorillamd) [inline src/JBController.sol:848]: nit

File: src/JBController.sol [hunk 4]
```diff
@@ -803,46 +808,98 @@ contract JBController is JBPermissioned, ERC2771Context, ERC165, IJBController,
...
+                        // have a token.
+                        IJBToken token = TOKENS.tokenOf(projectId);
+
+                        // Get a reference to the project's payment terminal that accepts the token.
+                        IJBTerminal terminal = token == IJBToken(address(0))
+                            ? IJBTerminal(address(0))
+                            : DIRECTORY.primaryTerminalOf(split.projectId, address(token));
+
+                        // If the paying project doesn't have a token or the receiving project isn't accepting the
+                        // token,
```
Comment (drgorillamd) [inline src/JBController.sol:858]: same, fmt nit

File: src/JBController.sol [hunk 5]
```diff
@@ -803,46 +808,98 @@ contract JBController is JBPermissioned, ERC2771Context, ERC165, IJBController,
...
+                            // Try to fulfill the payment
+                            try terminal.pay({
+                                projectId: split.projectId,
+                                token: address(token),
+                                amount: splitAmount,
+                                beneficiary: beneficiary,
+                                minReturnedTokens: 0,
+                                memo: "",
+                                metadata: metadata
+                            }) {} catch (bytes memory) {
```
Comment (drgorillamd) [inline src/JBController.sol:882]: we should at least emit an event with the reason

**Review by drgorillamd — APPROVED** [review]
smol comments, nothing blocking tho

File: src/JBController.sol [hunk 6]
```diff
@@ -803,46 +808,98 @@ contract JBController is JBPermissioned, ERC2771Context, ERC165, IJBController,
...
+                                // Transfer the tokens from this contract to the beneficiary.
+                                IERC20(address(token)).safeTransfer(beneficiary, splitAmount);
+                            }
+
+                            // Reset token approval of the terminal.
+                            IERC20(address(token)).approve(address(terminal), 0);
```
[the remark this answers is not in the document]
Comment (mejango) [reply src/JBController.sol:888]: @simplemachine92

**Pull request merged**